webknossos/tests/dataset/test_dataset_add_remote_mag_and_layer.py

```diff
@@ -0,0 +1,118 @@
+from pathlib import Path
+from tempfile import TemporaryDirectory
+from typing import Iterator
+
+import pytest
+
+import webknossos as wk
+from webknossos import Dataset, MagView
+from webknossos.utils import is_remote_path
+
+pytestmark = [pytest.mark.with_vcr]
+
+
+@pytest.fixture(scope="module")
+def sample_bbox() -> wk.BoundingBox:
+    return wk.BoundingBox((2807, 4352, 1794), (10, 10, 10))
+
+
+@pytest.fixture(scope="module")
+def sample_remote_dataset(sample_bbox: wk.BoundingBox) -> Iterator[wk.Dataset]:
+    url = "https://webknossos.org/datasets/scalable_minds/l4_sample_dev"
+    with TemporaryDirectory() as temp_dir:
+        yield wk.Dataset.download(url, path=Path(temp_dir) / "ds", bbox=sample_bbox)
+
+
+@pytest.fixture(scope="module")
+def sample_remote_mags() -> list[wk.MagView]:
+    mag_urls = [
+        "https://data-humerus.webknossos.org/data/zarr/scalable_minds/l4_sample_dev/color/1/",
+        "https://data-humerus.webknossos.org/data/zarr/scalable_minds/l4_sample_dev/color/2-2-1/",
+        "https://data-humerus.webknossos.org/data/zarr/scalable_minds/l4_sample_dev/color/4-4-2/",
+        "https://data-humerus.webknossos.org/data/zarr/scalable_minds/l4_sample_dev/segmentation/1/",
+        "https://data-humerus.webknossos.org/data/zarr/scalable_minds/l4_sample_dev/segmentation/2-2-1/",
+        "https://data-humerus.webknossos.org/data/zarr/scalable_minds/l4_sample_dev/segmentation/4-4-2/",
+    ]
+    mags = [MagView._ensure_mag_view(url) for url in mag_urls]
+    return mags
+
+
+@pytest.fixture(scope="module")
+def sample_remote_layer() -> list[wk.Layer]:
+    remote_dataset_url = "https://webknossos.org/datasets/scalable_minds/l4_sample_dev"
+    remote_dataset = Dataset.open_remote(remote_dataset_url)
+    return list(remote_dataset.layers.values())
+
+
+def test_add_remote_mags_from_mag_view(
+    sample_remote_mags: list[wk.MagView], sample_remote_dataset: wk.Dataset
+) -> None:
+    for remote_mag in sample_remote_mags:
+        mag_path = remote_mag.path
+        layer_type = remote_mag.layer.category
+        assert is_remote_path(mag_path), "Remote mag does not have remote path."
+        layer_name = f"test_remote_layer_{mag_path.parent.name}_{mag_path.name}_object"
+        new_layer = sample_remote_dataset.add_layer(
+            layer_name,
+            layer_type,
+            data_format=remote_mag.info.data_format,
+            dtype_per_channel=remote_mag.get_dtype(),
+        )
+        new_layer.add_remote_mag(remote_mag)
+        added_mag = sample_remote_dataset.layers[layer_name].mags[remote_mag.mag]
+        # checking whether the added_mag.path matches the mag_url with or without a trailing slash.
+        assert (
+            added_mag.path == mag_path or added_mag.path == mag_path.parent
+        ), "Added remote mag's path does not match remote path of mag added."
+
+
+"""
```

Comment:
Why are the two tests in this file commented out? If there's a valid reason for it, please add an explaining comment :)

Reply:
Thanks for finding this. The problem with the commented out tests is that they fail reliably and I don't know why. When I run the `test_add_remote_mags_from_path` test, loading the datasource properties json from `https://data-humerus.webknossos.org/data/zarr/scalable_minds/l4_sample_dev/datasource-properties.json` succeeds n times but the n+1st time fails. I added the following logging to `dataset.py _load_properties`:
```python
    def _load_properties(self) -> DatasetProperties:
        with (self.path / PROPERTIES_FILE_NAME).open(
            encoding="utf-8"
        ) as datasource_properties:
            contents = datasource_properties.read()
            print("file path")
            print(self.path / PROPERTIES_FILE_NAME)
            print("--------------------------")
            print("datasource_properties contents")
            print(contents)
            data = json.loads(contents)
        return dataset_converter.structure(data, DatasetProperties)
```
And the output of the failed test is:
```
file path
https://data-humerus.webknossos.org/data/zarr/scalable_minds/l4_sample_dev/datasource-properties.json
--------------------------
datasource_properties contents
{"id":{"name":"l4_sample_dev","team":"scalable_minds"},"dataLayers":[{"name":"color","category":"color","boundingBox":{"topLeft":[2679,4224,1719],"width":256,"height":256,"depth":150},"elementClass":"uint8","mags":[{"mag":[1,1,1],"axisOrder":{"x":1,"y":2,"z":3,"c":0}},{"mag":[2,2,1],"axisOrder":{"x":1,"y":2,"z":3,"c":0}},{"mag":[4,4,2],"axisOrder":{"x":1,"y":2,"z":3,"c":0}}],"defaultViewConfiguration":{"color":[255,219,219],"intensityRange":[70,180]},"numChannels":1,"dataFormat":"zarr"},{"name":"segmentation","boundingBox":{"topLeft":[2679,4224,1719],"width":256,"height":256,"depth":150},"elementClass":"uint32","mags":[{"mag":[1,1,1],"axisOrder":{"x":1,"y":2,"z":3,"c":0}},{"mag":[2,2,1],"axisOrder":{"x":1,"y":2,"z":3,"c":0}},{"mag":[4,4,2],"axisOrder":{"x":1,"y":2,"z":3,"c":0}}],"largestSegmentId":5238529,"numChannels":1,"dataFormat":"zarr","category":"segmentation"}],"scale":{"factor":[11.239999771118164,11.239999771118164,28],"unit":"nanometer"}}
file path
https://data-humerus.webknossos.org/data/zarr/scalable_minds/l4_sample_dev/datasource-properties.json
--------------------------
datasource_properties contents
{"id":{"name":"l4_sample_dev","team":"scalable_minds"},"dataLayers":[{"name":"color","category":"color","boundingBox":{"topLeft":[2679,4224,1719],"width":256,"height":256,"depth":150},"elementClass":"uint8","mags":[{"mag":[1,1,1],"axisOrder":{"x":1,"y":2,"z":3,"c":0}},{"mag":[2,2,1],"axisOrder":{"x":1,"y":2,"z":3,"c":0}},{"mag":[4,4,2],"axisOrder":{"x":1,"y":2,"z":3,"c":0}}],"defaultViewConfiguration":{"color":[255,219,219],"intensityRange":[70,180]},"numChannels":1,"dataFormat":"zarr"},{"name":"segmentation","boundingBox":{"topLeft":[2679,4224,1719],"width":256,"height":256,"depth":150},"elementClass":"uint32","mags":[{"mag":[1,1,1],"axisOrder":{"x":1,"y":2,"z":3,"c":0}},{"mag":[2,2,1],"axisOrder":{"x":1,"y":2,"z":3,"c":0}},{"mag":[4,4,2],"axisOrder":{"x":1,"y":2,"z":3,"c":0}}],"largestSegmentId":5238529,"numChannels":1,"dataFormat":"zarr","category":"segmentation"}],"scale":{"factor":[11.239999771118164,11.239999771118164,28],"unit":"nanometer"}}
file path
https://data-humerus.webknossos.org/data/zarr/scalable_minds/l4_sample_dev/datasource-properties.json
--------------------------
datasource_properties contents
{"id":{"name":"l4_sample_dev","team":"scalable_minds"},"dataLayers":[{"name":"color","category":"color","boundingBox":{"topLeft":[2679,4224,1719],"width":256,"height":256,"depth":150},"elementClass":"uint8","mags":[{"mag":[1,1,1],"axisOrder":{"x":1,"y":2,"z":3,"c":0}},{"mag":[2,2,1],"axisOrder":{"x":1,"y":2,"z":3,"c":0}},{"mag":[4,4,2],"axisOrder":{"x":1,"y":2,"z":3,"c":0}}],"defaultViewConfiguration":{"color":[255,219,219],"intensityRange":[70,180]},"numChannels":1,"dataFormat":"zarr"},{"name":"segmentation","boundingBox":{"topLeft":[2679,4224,1719],"width":256,"height":256,"depth":150},"elementClass":"uint32","mags":[{"mag":[1,1,1],"axisOrder":{"x":1,"y":2,"z":3,"c":0}},{"mag":[2,2,1],"axisOrder":{"x":1,"y":2,"z":3,"c":0}},{"mag":[4,4,2],"axisOrder":{"x":1,"y":2,"z":3,"c":0}}],"largestSegmentId":5238529,"numChannels":1,"dataFormat":"zarr","category":"segmentation"}],"scale":{"factor":[11.239999771118164,11.239999771118164,28],"unit":"nanometer"}}
file path
https://data-humerus.webknossos.org/data/zarr/scalable_minds/l4_sample_dev/datasource-properties.json
--------------------------
datasource_properties contents
{"id":{"name":"l4_sample_dev","team":"scalable_minds"},"dataLayers":[{"name":"color","category":"color","boundingBox":{"topLeft":[2679,4224,1719],"width":256,"height":256,"depth":150},"elementClass":"uint8","mags":[{"mag":[1,1,1],"axisOrder":{"x":1,"y":2,"z":3,"c":0}},{"mag":[2,2,1],"axisOrder":{"x":1,"y":2,"z":3,"c":0}},{"mag":[4,4,2],"axisOrder":{"x":1,"y":2,"z":3,"c":0}}],"defaultViewConfiguration":{"color":[255,219,219],"intensityRange":[70,180]},"numChannels":1,"dataFormat":"zarr"},{"name":"segmentation","boundingBox":{"topLeft":[2679,4224,1719],"width":256,"height":256,"depth":150},"elementClass":"uint32","mags":[{"mag":[1,1,1],"axisOrder":{"x":1,"y":2,"z":3,"c":0}},{"mag":[2,2,1],"axisOrder":{"x":1,"y":2,"z":3,"c":0}},{"mag":[4,4,2],"axisOrder":{"x":1,"y":2,"z":3,"c":0}}],"largestSegmentId":5238529,"numChannels":1,"dataFormat":"zarr","category":"segmentation"}],"scale":{"factor":[11.239999771118164,11.239999771118164,28],"unit":"nanometer"}}
file path
https://data-humerus.webknossos.org/data/zarr/scalable_minds/l4_sample_dev/datasource-properties.json
--------------------------
datasource_properties contents
{"id":{"name":"l4_sample_dev","team":"scalable_minds"},"dataLayers":[{"name":"color","category":"color","boundingBox":{"topLeft":[2679,4224,1719],"width":256,"height":256,"depth":150},"elementClass":"uint8","mags":[{"mag":[1,1,1],"axisOrder":{"x":1,"y":2,"z":3,"c":0}},{"mag":[2,2,1],"axisOrder":{"x":1,"y":2,"z":3,"c":0}},{"mag":[4,4,2],"axisOrder":{"x":1,"y":2,"z":3,"c":0}}],"defaultViewConfiguration":{"color":[255,219,219],"intensityRange":[70,180]},"numChannels":1,"dataFormat":"zarr"},{"name":"segmentation","boundingBox":{"topLeft":[2679,4224,1719],"width":256,"height":256,"depth":150},"elementClass":"uint32","mags":[{"mag":[1,1,1],"axisOrder":{"x":1,"y":2,"z":3,"c":0}},{"mag":[2,2,1],"axisOrder":{"x":1,"y":2,"z":3,"c":0}},{"mag":[4,4,2],"axisOrder":{"x":1,"y":2,"z":3,"c":0}}],"largestSegmentId":5238529,"numChannels":1,"dataFormat":"zarr","category":"segmentation"}],"scale":{"factor":[11.239999771118164,11.239999771118164,28],"unit":"nanometer"}}
file path
https://data-humerus.webknossos.org/data/zarr/scalable_minds/l4_sample_dev/datasource-properties.json
--------------------------
datasource_properties contents
{"id":{"name":"l4_sample_dev","team":"scalable_minds"},"dataLayers":[{"name":"color","category":"color","boundingBox":{"topLeft":[2679,4224,1719],"width":256,"height":256,"depth":150},"elementClass":"uint8","mags":[{"mag":[1,1,1],"axisOrder":{"x":1,"y":2,"z":3,"c":0}},{"mag":[2,2,1],"axisOrder":{"x":1,"y":2,"z":3,"c":0}},{"mag":[4,4,2],"axisOrder":{"x":1,"y":2,"z":3,"c":0}}],"defaultViewConfiguration":{"color":[255,219,219],"intensityRange":[70,180]},"numChannels":1,"dataFormat":"zarr"},{"name":"segmentation","boundingBox":{"topLeft":[2679,4224,1719],"width":256,"height":256,"depth":150},"elementClass":"uint32","mags":[{"mag":[1,1,1],"axisOrder":{"x":1,"y":2,"z":3,"c":0}},{"mag":[2,2,1],"axisOrder":{"x":1,"y":2,"z":3,"c":0}},{"mag":[4,4,2],"axisOrder":{"x":1,"y":2,"z":3,"c":0}}],"largestSegmentId":5238529,"numChannels":1,"dataFormat":"zarr","category":"segmentation"}],"scale":{"factor":[11.239999771118164,11.239999771118164,28],"unit":"nanometer"}}
file path
/tmp/tmp5izrse_t/ds/datasource-properties.json
--------------------------
datasource_properties contents
{
    "id": {
        ...
    },
    "scale": {
        ...
    },
    "dataLayers": []
}
file path
/tmp/tmp5izrse_t/ds/datasource-properties.json
--------------------------
datasource_properties contents
{
    "id": {
        ...
    },
    "scale": {
        ...
    },
    "dataLayers": []
}
file path
/tmp/tmp5izrse_t/ds/datasource-properties.json
--------------------------
datasource_properties contents
{
    "id": {
        ...
    },
    "scale": {
        ...
    },
    "dataLayers": [
        ...
    ]
}
file path
/tmp/tmp5izrse_t/ds/datasource-properties.json
--------------------------
datasource_properties contents
{
    "id": {
        ...
    },
    "scale": {
        ...
    },
    "dataLayers": [
        ...
    ]
}
file path
/tmp/tmp5izrse_t/ds/datasource-properties.json
--------------------------
datasource_properties contents
{
    "id": {
        ...
    },
    "scale": {
        ...
    },
    "dataLayers": [
        ...
    ]
}
Downloading layer=color mag=1 ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00
file path
/tmp/tmp5izrse_t/ds/datasource-properties.json
--------------------------
datasource_properties contents
{
    "id": {
        ...
    },
    "scale": {
        ...
    },
    "dataLayers": [
        ...
    ]
}
Downloading layer=color mag=2-2-1 ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00
file path
/tmp/tmp5izrse_t/ds/datasource-properties.json
--------------------------
datasource_properties contents
{
    "id": {
        ...
    },
    "scale": {
        ...
    },
    "dataLayers": [
        ...
    ]
}
Downloading layer=color mag=4-4-2 ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00
file path
/tmp/tmp5izrse_t/ds/datasource-properties.json
--------------------------
datasource_properties contents
{
    "id": {
        ...
    },
    "scale": {
        ...
    },
    "dataLayers": [
        ...
    ]
}
file path
/tmp/tmp5izrse_t/ds/datasource-properties.json
--------------------------
datasource_properties contents
{
    "id": {
        ...
    },
    "scale": {
        ...
    },
    "dataLayers": [
        ...
    ]
}
file path
/tmp/tmp5izrse_t/ds/datasource-properties.json
--------------------------
datasource_properties contents
{
    "id": {
        ...
    },
    "scale": {
        ...
    },
    "dataLayers": [
        ...
    ]
}
Downloading layer=segmentation mag=1 ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00
file path
/tmp/tmp5izrse_t/ds/datasource-properties.json
--------------------------
datasource_properties contents
{
    "id": {
        ...
    },
    "scale": {
        ...
    },
    "dataLayers": [
        ...
    ]
}
Downloading layer=segmentation mag=2-2-1 ━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00
file path
/tmp/tmp5izrse_t/ds/datasource-properties.json
--------------------------
datasource_properties contents
{
    "id": {
        ...
    },
    "scale": {
        ...
    },
    "dataLayers": [
        ...
    ]
}
Downloading layer=segmentation mag=4-4-2 ━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00
----------------------------------------------------------------------------------------------- Captured stdout call ------------------------------------------------------------------------------------------------
file path
/tmp/tmp5izrse_t/ds/datasource-properties.json
--------------------------
datasource_properties contents
{
    "id": {
        ...
    },
    "scale": {
        ...
    },
    "dataLayers": [
        ...
    ]
}
file path
https://data-humerus.webknossos.org/data/zarr/scalable_minds/l4_sample_dev/datasource-properties.json

--------------------------
datasource_properties contents


```

As you can see the output shows the first requests to `https://data-humerus.webknossos.org/data/zarr/scalable_minds/l4_sample_dev/datasource-properties.json` result in a valid json result but the last logged access to the file just returns an empty string :/


Very random thought @fm3 is maybe some kind of ddos protection / rate limiting active for `data-humerus.webknossos.org` that might explain this behaviour :thinking:?

Reply:
But nevertheless, I remove the commented out tests as they can be recovered from the commit history of this branch

Reply:
Shouldn’t think that there is such a rate limiting in place there. Could you also print the request reply etc? This is opened using upath, right? Maybe it doesn’t handle the requests correctly?
Also, why does the test have to open the json so many times?